### PR TITLE
[frontend] refactored list_users method

### PR DESCRIPTION
**Pull request by vpereira**
just because why not. I saw the code today, I couldnt sleep without try to change it :) 

other solution would be using https://dev.mysql.com/doc/refman/5.7/en/regexp.html#operator_regexp but it would change the functionality a little..

### CODE REVIEW

**Comment by mdeniz** [comment]
Nice that you started this :smile: 

Can we not stop refactoring the inside of the `list_users` but also thinking about moving this to the proper place? I mean the model or a decorator instead of letting this live in the controller.

**Comment by vpereira** [comment]
@mdeniz Hey Moisés thank you, Probably I did something wrong in the spec files, don't know why I wasnt able to create the user as I want. Can you spot it any issue?

File: src/api/app/models/user.rb [hunk 1]
```diff
@@ -127,6 +128,18 @@ def create_home_project
     self.login_failure_count = 0 if login_failure_count.nil?
   end
 
+  # autocomplete method
+  # return either an array of users or an array of hashes
+  def self.autocomplete(prefix = nil, is_token = false)
+    User.where('login LIKE ?', "#{prefix}%").pluck(:login).collect do |user|
```
Comment (mdeniz) [inline src/api/app/models/user.rb]: You don't need to put the `User` class in front as you are already in the `User` model

Comment (mdeniz) [reply src/api/app/models/user.rb]: What if you create a scope out of `where('login LIKE ?', "#{prefix}%")`

File: src/api/spec/models/user_spec.rb [hunk 2]
```diff
@@ -870,4 +870,19 @@
       end
     end
   end
+
+  describe '#autocomplete' do
+    let(:user) { create(:confirmed_user, login: 'foobar') }
```
Comment (mdeniz) [inline src/api/spec/models/user_spec.rb]: For explicitly create a user you need to either call it somewhere (normally in a `before` block or in the it example) or just use the `let!`method instead that calls the creation in every single test.

BTW, I would love to see more users created with different names just to be sure that the matched ones are retrieved and not the rest. You don't only need to test that the output format is ok, you need to be sure that this finds the right objects.

File: src/api/spec/models/user_spec.rb [hunk 3]
```diff
@@ -870,4 +870,19 @@
...
+
+    context 'autocomplete' do
+      subject { User.autocomplete('foo') }
+
+      it { expect(subject).to include('foobar') }
+    end
+    context 'tokens' do
```
Comment (mdeniz) [inline src/api/spec/models/user_spec.rb]: You need a space here

File: src/api/spec/models/user_spec.rb [hunk 4]
```diff
@@ -1,5 +1,5 @@
 require 'rails_helper'
-
+# rubocop:disable Metrics/BlockLength
```
Comment (mdeniz) [inline src/api/spec/models/user_spec.rb]: We usually don't do this, we normally add this file to the ignored files for that cop.

File: src/api/app/models/user.rb [hunk 5]
```diff
@@ -127,6 +128,18 @@ def create_home_project
     self.login_failure_count = 0 if login_failure_count.nil?
   end
 
+  # autocomplete method
+  # return either an array of users or an array of hashes
+  def self.autocomplete(prefix = nil, is_token = false)
```
Comment (mdeniz) [inline src/api/app/models/user.rb]: I would prefer to have a separate method for autocomplete tokens, that way you don't have to check it for every single item neither to pass a param for it and the collect would be super easy in each case. At the same time you can use each method in the corresponding action of the controller.

File: src/api/spec/models/user_spec.rb (hunk 4 above)
Comment (vpereira) [reply src/api/spec/models/user_spec.rb]: ok, I added 
```
Metrics/BlockLength:
  Exclude:
    - "**/*_spec.rb"

```

and then rubocop complained that we have alread it on:
```
src/api/spec/controllers/webui/package_controller_spec.rb:# rubocop:disable Metrics/BlockLength
src/api/spec/controllers/webui/project_controller_spec.rb:# rubocop:disable Metrics/BlockLength
```

thats why I did it online, but I can put it global as I did. It make sense for Rspec which are mainly blocks

File: src/api/spec/models/user_spec.rb [hunk 6]
```diff
@@ -870,4 +870,22 @@
...
   end
+
+  describe '#autocomplete' do
+    let!(:foobar) { create(:confirmed_user, login: 'foobar') }
+    let!(:fobaz) { create(:confirmed_user, login: 'fobaz') }
+
+    context 'autocomplete' do
+      subject { User.autocomplete('foo') }
+
+      it { expect(subject).to include('foobar') }
```
Comment (mdeniz) [inline src/api/spec/models/user_spec.rb]: In the `include` cases I would prefer to use `match_array` that checks exactly what values the array have, because you want to know which values are retrieved, not only that the value you are expecting is there.

File: src/api/spec/models/user_spec.rb [hunk 7]
```diff
@@ -870,4 +870,22 @@
...
+      it { expect(subject).not_to include('fozbar') }
+    end
+    context 'tokens' do
```
Comment (mdeniz) [inline src/api/spec/models/user_spec.rb]: You still need a blank line between contexts

File: src/api/spec/models/user_spec.rb (hunk 4 above)
Comment (Ana06) [reply src/api/spec/models/user_spec.rb]: @mdeniz 

> We usually don't do this, we normally add this file to the ignored files for that cop.

We actually do it... It is in several places in the tests... but we shouldn't :see_no_evil: :joy:

Comment (vpereira) [reply src/api/spec/models/user_spec.rb]:
```
grep -r "# rubocop:disable " * | wc -l
74
```

File: src/api/app/models/user.rb (hunk 5 above)
Comment (ChrisBr) [reply src/api/app/models/user.rb]: I also don't like to check the is_token in the loop as it won't change. Also in case of is_token = false you iterate over the items instead of just returning them.
What about:

```
result =  where('login LIKE ?', "#{prefix}%").pluck(:login)
is_token ? result.collect { |user| { name: user } }: result
```

File: src/api/app/models/user.rb [hunk 8]
```diff
@@ -95,6 +95,7 @@ class User < ApplicationRecord
   validates :password, confirmation: true, allow_blank: true
 
   after_create :create_home_project
+
```
Comment (Ana06) [inline src/api/app/models/user.rb]: unrelated change to the PR :confounded:

File: src/api/app/models/user.rb [hunk 9]
```diff
@@ -127,6 +128,18 @@ def create_home_project
     self.login_failure_count = 0 if login_failure_count.nil?
   end
 
+  # autocomplete method
```
Comment (Ana06) [inline src/api/app/models/user.rb]: isn't that clear with the name of the method? :joy:

File: src/api/spec/models/user_spec.rb (hunk 4 above)
Comment (mdeniz) [reply src/api/spec/models/user_spec.rb]: I know this exists, but we had agreed in not doing this anymore...

File: src/api/app/models/user.rb (hunk 5 above)
Comment (Ana06) [inline src/api/app/models/user.rb]: Now that you are refactoring, what about having two autocomplete method instead of the `is_token` parameter? It only makes more complicate to understand it. :thinking:

File: src/api/spec/models/user_spec.rb [hunk 10]
```diff
@@ -870,4 +870,23 @@
       end
     end
   end
+
+  describe '#autocomplete' do
+    let!(:foobar) { create(:confirmed_user, login: 'foobar') }
```
Comment (Ana06) [inline src/api/spec/models/user_spec.rb]: [nitpick] what about using something that seems to be a login? like a real name? :smile:

File: src/api/spec/models/user_spec.rb (hunk 6 above)
Comment (Ana06) [reply src/api/spec/models/user_spec.rb]: :+1: to  @mdeniz, then you can remove the `not_to include` case. :tada:

File: src/api/app/models/user.rb [hunk 11]
```diff
@@ -63,6 +63,7 @@ class User < ApplicationRecord
   has_many :rss_feed_items, -> { order(created_at: :desc) }, class_name: 'Notification::RssFeedItem', as: :subscriber, dependent: :destroy
 
   scope :all_without_nobody, -> { where('login != ?', nobody_login) }
+  scope :autocomplete_login, ->(prefix) { where('login LIKE ?', "#{prefix}%") if prefix.present? }
```
Comment (mdeniz) [inline src/api/app/models/user.rb]: Mmmm, why do you check that the prefix is present?
And if this is the case and is not present? what would happen?

File: src/api/Gemfile.lock [hunk 12]
```diff
@@ -328,7 +328,7 @@ GEM
     runcom (2.0.1)
       refinements (~> 5.0)
     safe_yaml (1.0.4)
-    sanitize (4.6.0)
+    sanitize (4.6.4)
```
[the remark this answers is not in the document]
Comment (vpereira) [reply src/api/Gemfile.lock]: hakiri tests were failing and breaking my "ALL GREEN" apart :)

File: src/api/spec/models/user_spec.rb (hunk 10 above)
Comment (vpereira) [reply src/api/spec/models/user_spec.rb]: because its a login and not real name, even described in RFCs https://www.ietf.org/rfc/rfc3092.txt :)

File: src/api/app/models/user.rb (hunk 11 above)
Comment (vpereira) [reply src/api/app/models/user.rb]: if the scope its called without any parameter, you get an ArgumentError error
if you pass nil or ```''``` and you dont check the prefix, then the '%' will be passed to the query forming a query like ```SELECT `users`.`login` FROM `users` WHERE (login LIKE '%')``` and probably returning all users?

Comment (vpereira) [reply src/api/app/models/user.rb]: I'm adding a test case for that. I think that is probably the issue with using scope here. Normally scope behaves like a class method returning either scoped or all. So more or less like:

```
def self.autocomplete(prefix)
   if prefix.present?
    where('login LIKE ?', "#{prefix}%")
  else
    all
  end
end
```

Comment (mdeniz) [reply src/api/app/models/user.rb]: But the scope shouldn't care too much about the prefix length... I would move this control either to the #autocomplete method or to the controller side.

File: src/api/app/models/user.rb [hunk 13]
```diff
@@ -63,6 +63,7 @@ class User < ApplicationRecord
   has_many :rss_feed_items, -> { order(created_at: :desc) }, class_name: 'Notification::RssFeedItem', as: :subscriber, dependent: :destroy
 
   scope :all_without_nobody, -> { where('login != ?', nobody_login) }
+  scope :autocomplete_login, ->(prefix) { prefix.present? ? where('login LIKE ?', "#{prefix}%") : User.none }
```
Comment (mdeniz) [inline src/api/app/models/user.rb]: Can you maybe change the name of the scope to something like "with_login_prefix" ?? that way when you use it it will be something like "User.with_login_prefix('foo')"

File: src/api/app/models/user.rb [hunk 14]
```diff
@@ -129,6 +130,22 @@ def create_home_project
     self.login_failure_count = 0 if login_failure_count.nil?
   end
 
+  def self.autocomplete_login(prefix = '')
```
Comment (mdeniz) [inline src/api/app/models/user.rb]: For this you can just use an `alias`

File: src/api/app/controllers/webui/user_controller.rb [hunk 15]
```diff
@@ -211,27 +211,12 @@ def change_password
...
-    users = User.arel_table
-    User.where(users[:login].matches("#{prefix}%")).pluck(:login).each do |user|
-      if hash
-        names << { 'name' => user }
-      else
-        names << user
-      end
-    end
-    names
+    render json: User.autocomplete(params[:q], true)
```
Comment (mdeniz) [inline src/api/app/controllers/webui/user_controller.rb]: This won't work, now the method is `autocomplete_token`

File: src/api/app/controllers/webui/user_controller.rb [hunk 16]
```diff
@@ -211,27 +211,12 @@ def change_password
 
   def autocomplete
     required_parameters :term
-    render json: list_users(params[:term])
+    render json: User.autocomplete(params[:term])
```
Comment (mdeniz) [inline src/api/app/controllers/webui/user_controller.rb]: This won't work, now the method is `autocomplete_login`

File: src/api/app/models/user.rb (hunk 14 above)
Comment (Ana06) [reply src/api/app/models/user.rb]: why is actually this method needed? why not just using `autocomplete_backend` directly? as @mdeniz it is just the same thing with different name.... :confounded:

File: src/api/app/models/user.rb [hunk 17]
```diff
@@ -129,6 +130,22 @@ def create_home_project
...
+
+  def self.autocomplete_token(prefix = '')
+    autocomplete_backend(prefix).collect { |user| { name: user } }
+  end
+
+  def self.autocomplete_backend(prefix = '')
+    if prefix.present?
+      with_login_prefix(prefix).pluck(:login)
+    else
+      User.none
```
Comment (Ana06) [inline src/api/app/models/user.rb]: this return a relation, while in the if part you return an array... I would return `[]` in this case for consistence instead.

Even more, I would write this like:

```ruby
return [] unless prefix.present?
with_login_prefix(prefix).pluck(:login)

Comment (mdeniz) [reply src/api/app/models/user.rb]: I already pointed to this in a private conversation :1st_place_medal:

File: .rubocop.yml [hunk 18]
```diff
@@ -68,6 +68,11 @@ Style/WordArray:
 
 ##################### Metrics ##################################
 
+# Block Length check disabled just for specs
+Metrics/BlockLength:
```
[the remark this answers is not in the document]
Comment (bgeuken) [reply .rubocop.yml]: @Ana06 Actually I just [did](https://github.com/openSUSE/open-build-service/pull/4681/files#diff-74423563ec7c74f4cb7f74689b49e7c1R76) it :laughing: . And even reduced it by moving some tests into another spec :bowtie:

**Comment by bgeuken** [comment]
Nice work :bouquet: LGTM, except the things the others already pointed out

**Comment by vpereira** [comment]
Could someone take this PR over? I won't have time to touch it in the next two weeks and I don't want to let it open.. Thanks!

**Comment by bgeuken** [comment]
> vpereira commented 4 days ago
> Could someone take this PR over? I won't have time to touch it in the next two weeks and I don't want to let it open.. Thanks!

Okay. I'll take care of it:+1:
